src/main/java/org/testng/internal/ExitCode.java

```diff
@@ -0,0 +1,148 @@
+
+        static Status parse(int status) {
+            for (Status each : values()) {
+                if ((each.ordinal() + 1) == status) {
+                    return each;
+                }
+            }
+            return null;
+        }
+    }
+
+    public static boolean hasFailureWithinSuccessPercentage(int retCode) {
+        Status status = Status.parse(retCode);
+        if (status == null) {
+            return false;
+        }
+        switch (status) {
+            case FAILEDWITHINSUCCESS:
+            case FAILEDWITHINSUCCESS_SUCCESS:
+            case FAILEDWITHINSUCCESS_FAILED:
+            case FAILEDWITHINSUCCESS_FAILED_SUCCESS:
+            case FAILEDWITHINSUCCESS_SKIPPED:
+            case FAILEDWITHINSUCCESS_SKIPPED_SUCCESS:
+            case FAILEDWITHINSUCCESS_SKIPPED_FAILED:
+            case FAILEDWITHINSUCCESS_SKIPPED_FAILED_SUCCESS:
+                return true;
+            default :
+                return false;
+        }
+    }
+
+    public static boolean hasSkipped(int retCode) {
+        Status status = Status.parse(retCode);
+        if (status == null) {
+            return false;
+        }
+        switch (status) {
+            case SKIPPED:
+            case SKIPPED_PASSED:
+            case SKIPPED_FAILED:
+            case SKIPPED_FAILED_PASSED:
+            case FAILEDWITHINSUCCESS_SKIPPED:
+            case FAILEDWITHINSUCCESS_SKIPPED_SUCCESS:
+            case FAILEDWITHINSUCCESS_SKIPPED_FAILED:
+            case FAILEDWITHINSUCCESS_SKIPPED_FAILED_SUCCESS:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    public static boolean hasFailure(int retCode) {
+        Status status = Status.parse(retCode);
+        if (status == null) {
+            return false;
+        }
+        switch (status) {
+            case PASSED_FAILED:
+            case SKIPPED_FAILED:
+            case SKIPPED_FAILED_PASSED:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private BitSet initial = new BitSet(4);
+
+    void computeAndUpdate(ITestContext context) {
+        boolean passed = (context.getPassedTests().size() != 0) ||
+                (context.getPassedConfigurations().size() != 0);
+        initial.set(0, passed);
+        boolean failed = (context.getFailedTests().size() != 0) ||
+                (context.getFailedConfigurations().size() != 0);
+        initial.set(1, failed);
+        boolean skipped = (context.getSkippedTests().size() != 0) ||
+                (context.getSkippedConfigurations().size() != 0);
+        initial.set(2, skipped);
+        boolean failedWithinSuccessPercentage = (context.getFailedButWithinSuccessPercentageTests().size() != 0);
+        initial.set(3, failedWithinSuccessPercentage);
+    }
+
+    int getExitCode() {
+        int size = initial.length() - 1;
+        int value = 0;
+        for (int i = size; i >= 0; i--) {
```

Comment:
this implementation looks easier: https://stackoverflow.com/a/4874090/4234729

Reply:
I did see it. But I wasn't comfortable adopting it, because I didn't understand what the code was doing. I think the last line viz., `result &= Integer.MAX_VALUE;` is not required. I will go ahead and adopt it.